### ISSUE TITLE
More compatibilty in clause references

### ISSUE THREAD

**Issue by jburse** [jburse/jekejeke-devel#114]
We could adopt number_of_clauses/1 predicate property
and nth_clause/3 builtin. To code for example:
```
random_word(W) :-
    predicate_propery(word(_), number_of_clauses(N)),
    random_between(1, N, Rnd),
    nth_clause(word(_), Rnd, Ref),
    clause(word(W), true, Ref).
```
But are feasible in our system, since we have ropes. A rope
allows size and random access. In some places, we use call_nth/2, 
maybe nth_clause/3 could be a replacement for some 

of these calls? Unlikely, I guess the crucial use case is
random access, where the index is given, and not some
enumeration of clauses.

**Comment by jburse** [comment]
Some Prolog systems have next_clause/2 for clause references.
We also saw that some Prolog systems have assertz/2 returning
a clause reference.

But the original clause reference API is anyway incompatible with
our reference API. It seems that the Key-Term pair cannot be
used as a substitute to Head-Clause, since the variables do not

communicate. This is seen here in SWI-Prolog, the variable
in the foo/1 does not anymore communicate with the variable
in bar/2 after retrieval:
```
?- recordz(foo(X), bar(X,Y), R), recorded(P, Q, R).
R = <record>(0x600001375860),
P = foo(_8896),
Q = bar(_8900, _8902).
```
But most of our strategy is that clause properties wander from
the clause skeleton to the instanted term, like the variable names,
so that this can be used by the debugger, that only sees

the goals in the body of a clause. So that we don't need clause
property, since we have callable_property/2 now. We could
therefore remove our ref_property/2, set_ref_property/2
and reset_ref_property/2, sinced it was never used.

Maybe we could introduced erased/1 test predicate, which 
we saw somewhere else, but not in SWI-Prolog. Or keep
ref_property/2 to determine erased state?
